pkg/pillar/cmd/downloader/downloader.go

```diff
@@ -1176,6 +1183,166 @@ func doSftp(ctx *downloaderContext, status *types.DownloaderStatus,
+			break
+		}
+	}
+	log.Infof("rktFetch - imageId: %s\n", imageId)
+	if imageId == "" {
+		errMsg := "rkt fetch: Can't find imageId.\n Fetch Output: " +
+			outputStr
+		return "", errors.New(errMsg)
+	}
+
+	// XXX:FIXME - we should run "rkt image ls" and verify image fetch
+	// went thru without errors.
+	return imageId, nil
+}
+
+func rktAuthFilename(appName string) string {
+	return persistRktLocalConfigAuthDir + "/rktAuth" + appName + ".json"
+}
+
+func rktCreateAuthFile(config *types.DownloaderConfig) (string, error) {
+
+	if len(strings.TrimSpace(config.ApiKey)) == 0 {
+		log.Debugf("rktCreateAuthFile: empty username. Skipping AuthFile")
+		return "", nil
+	}
+
+	// XXX:FIXME - Why should this directory be created again
+	// when it was created in the first place in device-steps.sh
+	err := os.MkdirAll(persistRktLocalConfigAuthDir, 0755)
+	if err != nil {
+		log.Errorf("rktCreateAuthFile: empty username. Skipping AuthFile")
+		return "", errors.New(fmt.Sprintf("Failed create dir %s, "+
+			"err: %+v\n", persistRktLocalConfigAuthDir, err))
+	}
+
+	filename := rktAuthFilename(config.Safename)
+
+	rktAuth := types.RktAuthInfo{
+		RktKind:    "dockerAuth",
+		RktVersion: "v1",
+		Registries: []string{"registry-1.docker.io"},
+		Credentials: types.RktCredentials{
+			User:     config.ApiKey,
+			Password: config.Password,
+		},
+	}
+	log.Infof("rktCreateAuthFile: created Auth file %s\n"+
+		"rktAuth: %+v\n", filename, rktAuth)
+
+	file, err := json.MarshalIndent(rktAuth, "", " ")
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Failed convert rktAuth to json"+
+			"err: %+v\n", err))
+	}
+	err = ioutil.WriteFile(filename, file, 0644)
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Failed to create Auth file for"+
+			"rkt fetch: %+v\n", err))
+	}
+	return filename, nil
+}
+
+func rktFetchContainerImage(ctx *downloaderContext, key string,
+	config types.DownloaderConfig, status *types.DownloaderStatus) error {
+	// update status to DOWNLOAD STARTED
+	status.State = types.DOWNLOAD_STARTED
+	publishDownloaderStatus(ctx, status)
+
+	imageId := ""
+	// Save credentials to Auth file
+	log.Infof("rktFetchContainerImage: fetch  <%s>\n", config.DownloadURL)
+	authFile, err := rktCreateAuthFile(&config)
+	if err == nil {
+		log.Debugf("rktFetchContainerImage: authFile: %s\n", authFile)
+		// We should really move to have per-fetch directory..
+		localConfigDir := persistRktLocalConfigDir
+		if len(authFile) == 0 {
+			localConfigDir = ""
+			log.Infof("rktFetchContainerImage: no Auth File")
+		}
+		imageId, err = rktFetch(config.DownloadURL, localConfigDir)
+	} else {
+		log.Errorf("rktCreateAuthFile Failed. err: %+v", err)
+	}
+
+	if err != nil {
+		log.Errorf("rktFetchContainerImage: fetch  Failed. url:%s, "+
+			"authFile: %s, Err: %+v\n", config.DownloadURL, authFile, err)
+		status.PendingAdd = false
+		status.Size = 0
+		status.LastErr = fmt.Sprintf("%v", err)
+		status.LastErrTime = time.Now()
+		status.RetryCount += 1
```

Comment:
status.State doesn't change here with download failed. How many time RetryCount is going to go up?

Reply:
Right now there is no retry..

Currently, there is no DOWNLOAD_ERROR state. status.State is looked at only if there is no error.